### PR TITLE
Implement Min P as a sampler option in HF loaders

### PR DESCRIPTION
**Pull request by kalomaze**
## Checklist:

- [✔️] I have read the [Contributing guidelines](https://github.com/oobabooga/text-generation-webui/wiki/Contributing-guidelines).

A backport of the Min P sampler option that was merged into [llama.cpp recently.](https://github.com/ggerganov/llama.cpp/pull/3841) I think this is a suitable alternative to accomodate for the flaws of both Top P (Top P doesn't work well if there's not good concentration of high probability for top tokens) and Top K (can limit diversity unnecessarily), and it has gained a very positive reception.

image

### CODE REVIEW

**Comment by kalomaze** [comment]
Technically the min_tokens_to_keep is not being properly followed here which probably needs to be changed before it gets merged.

**Comment by kalomaze** [comment]
Should be good to merge now.

**Comment by oobabooga** [comment]
I was about to add min p manually since it looks cool. Thanks for the PR.

About the order, temperature does come before top p/top k in the transformers library:

https://github.com/huggingface/transformers/blob/acc394c4f5e1283c19783581790b3dc3105a3697/src/transformers/generation/utils.py#L825

I couldn't find any information as to how OpenAI does it.

Personally, I think that it makes sense to use temperature first.

**Comment by kalomaze** [comment]
> I was about to add min p manually since it looks cool. Thanks for the PR.
> 
> About the order, temperature does come before top p/top k in the transformers library:
> 
> https://github.com/huggingface/transformers/blob/acc394c4f5e1283c19783581790b3dc3105a3697/src/transformers/generation/utils.py#L825
> 
> I couldn't find any information as to how OpenAI does it.
> 
> Personally, I think that it makes sense to use temperature first.

I can see arguments for both applying it last like how it is done in llama.cpp, and applying it like how it's done 'officially', so I think having this configurable / customizable to the end user would be helpful. Is that possible without changing how the transformers library functions?

**Comment by kalomaze** [comment]
Changing how the sampler order logic works is out of scope for this PR though, so I'm ready for this to be merged in the meantime

**Comment by BadisG** [comment]
I think you should put the default min_p at 0 instead of 1, having it at 1 makes everything else useless because it's the equivalent to top_k = 1

**Comment by kalomaze** [comment]
> I think you should put the default min_p at 0 instead of 1, having it at 1 makes everything else useless because it's the equivalent to top_k = 1

Is it not 0 by default? I thought I set it to that

**Comment by oobabooga** [comment]
Should be fixed now https://github.com/oobabooga/text-generation-webui/commit/4537853e2c1760a00bd278fb97f26f4c4fb149cb

**Comment by BadisG** [comment]
It's still at min_p = 1 by default for me

**Comment by oobabooga** [comment]
I messed up again, should be good after https://github.com/oobabooga/text-generation-webui/commit/7f9c1cbb30b13a313923d994b34ce175c522f9d8

**Comment by BadisG** [comment]
All right it's working now, thanks!

**Comment by BadisG** [comment]
I've also noticed that min_p and presence_penalty don't mix well, from time to time my generation stops abruptly because min_p removes some tokens and presence_penalty then removes the remaining tokens, leaving nothing to go on. I think this can be fixed by reversing the order of the samplers, presence_penalty should be applied before min_p. Maybe we should have the choice of sampler order, that would be a good feature to add, a bit like KoboldAI actually.

**Comment by kalomaze** [comment]
> I've also noticed that min_p and presence_penalty don't mix well, from time to time my generation stops abruptly because min_p removes some tokens and presence_penalty then removes the remaining tokens, leaving nothing to go on. I think this can be fixed by reversing the order of the samplers, presence_penalty should be applied before min_p. Maybe we should have the choice of sampler order, that would be a good feature to add, a bit like KoboldAI actually.

Presence penalty... removes the tokens? I thought the rep pen (and associated version of it) are biasing the logits and not actually truncating them based on a certain criteria, that's weird.
